libgadget/cooling_qso_lightup.c

```diff
@@ -0,0 +1,613 @@
+/** This file implements the quasar heating model of Upton-Sanderbeck et al 2019 (in prep).
+ *  A black hole particle with the right mass is chosen at random and an ionizing bubble created around it.
+ *  Particles within that bubble are marked as ionized and heated according to emissivity.
+ *  New bubbles are created until the total HeIII fraction matches the value in the external table.
+ *  There is also a uniform background heating rate for long mean-free-path photons with very high energies.
+ *  This heating is added only to not-yet-ionized particles, and is done in cooling_rates.c. Whatever UVB you
+ *  use should not include these photons.
+ *
+ * The text file contains the reionization history and is generated from various physical processes
+ * implemented in a python file. The text file fixes the end of helium reionization (which is reasonably well-known).
+ * The code has the start time of reionization as a free parameter: if this start of reionization is late then there
+ * will be a strong sudden burst of heating.
+ *
+ * (Loosely) based on lightup_QSOs.py from https://github.com/uptonsanderbeck/helium_reionization
+ * HeII_heating.py contains the details of the reionization history and how it is generated.
+ *
+ * This code should run only during a PM timestep, when all particles are active
+ * We lose time resolution but I cannot think of another way to ensure cooling is modelled correctly.
+ */
+
+/*
+ * 1. Use a BH accretion rate threshold rather than a BH mass threshold.
+ * 2. Bubble size correlates with black hole mass.
+ *
+ * 3. Add a switch to place bubbles at random in massive halos if there are no black holes.
+ * 4. Bubble size correlates with halo mass.
+ */
+
+#include <math.h>
+#include <mpi.h>
+#include <string.h>
+#include <gsl/gsl_interp.h>
+#include "cooling_qso_lightup.h"
```

Comment:
Move per-module header to the end of includes, if possible.
If the move causes an error then we may have early signs of circular dependency.